### DISCUSSION TITLE
Final deliverables for each lens

### DISCUSSION THREAD

**Discussion by ajshajib** [AstroBridge/BDLensing#227] [Announcements]
To close the final deliverables issue for each lens system, the following products need to be provided:

1. A notebook that estimates $R_{10}$ and $\Sigma_{10}$ for the lens system, with name "estimate_sigma_10_system_name.ipynb".
1. The final modeling run needs to have "supersampling_factor" = 3 in `kwargs_numerics`, so set "supersampling_factor"  in `kwargs_numerics` to 3 and extend your chain from the last point until it converges again.
2. The `output` variable that is saved in "*_mcmc_output.joblib" file needs to be in this format: `[kwargs_result_mcmc, multi_band_list_out, kwargs_fixed_out, shortened_samples_mcmc, param_mcmc]`. The `shortened_samples_mcmc` must have a shape (1000, number of parameters) and only include values from the MCMC chain after convergence (and after thinning it out). You should update the modeling notebook to change the saved file and also to save the additional file described in the next point.
3. A YAML file with the name "system_name_point_estimates.yml" that provides the median, upper $1\sigma$ uncertainty (84th percentile $-$ median), and lower $1\sigma$ uncertainty (median $-$ 16th percentile) for these parameters: `theta_E`, `gamma`, `q_mass`, `phi_mass`, `center_x_mass`, `center_y_mass`,  `gamma_shear`, `phi_shear`, `R_eff_light`, `q_light`, `phi_light`, `center_x_light`, `center_y_light`, `Sigma_10`, `Sigma_10_flux_selected`, `Sigma_20`, `Sigma_20_flux_selected`. Use the `yaml` package to save an appropriately constructed dictionary in this YAML file. For example, the file should have the parameter values this way:
```
theta_E: [median, upper, lower]
gamma: [median, upper, lower]
e1_mass: ...
```

**Comment by ajshajib** [comment]
I don't see any PR created yet to close the "final deliverable" issues. This should be a quick PR without requiring too much work, except for probably the first system to make the $\Sigma_{10}$ measurement notebook. 

Please create the PRs for the seven systems, for which MCMC is done, with the goal to merge them by **Nov 30**. This should be the higher priority now over working on modeling more systems, as these deliverables are required to complete a full draft of the paper.

@Rajit13 @rafee-adnan @aaimtiaz @mamun-nahid @JobairHasan @SulymanHossain

**Comment by ajshajib** [comment]
I noticed that most people are using "supersampling_factor" = 1, which is fine for preliminary runs, but for the final run to put on the paper, we need to run with "supersampling_factor" = 3. I have added this point to the above list as point 2. One option to update the settings would be

```
kwargs_numerics_hires = copy.deepcopy(kwargs_numerics)
kwargs_numerics_hires["supersampling_factor"] = 3

fitting_kwargs_list = [
    ["update_settings", {"kwargs_numerics": kwargs_numerics_hires}],
    # add fitting step to extend MCMC here
]
```

**Reply by aaimtiaz** [reply]
FittingSequence.update_settings() got an unexpected keyword argument 'kwargs_numerics'
There is no argument named kwargs_numerics.

![image](https://github.com/ajshajib/BDLensing/assets/78560474/e5cfd587-a5c9-412b-a16e-64adcd0776f3)

**Comment by aaimtiaz** [comment]
I have created the yaml file in this way:
```
import yaml
import numpy as np

#list of parameters to be included in the yaml file
list_params = ['theta_E_lens0', 'gamma_lens0',
               'e1_lens0', 'e2_lens0',
               'center_x_lens0', 'center_y_lens0',
               'gamma_ext_lens1', 'psi_ext_lens1',
               'theta_E_lens2', 'R_sersic_lens_light0',
               'e1_lens_light0', 'e2_lens_light0',
               'center_x_lens_light0', 'center_y_lens_light0',
               'R_sersic_lens_light1', 'e1_lens_light1',
               'e2_lens_light1', 'R_sersic_lens_light2',
               'e1_lens_light2', 'e2_lens_light2',
               'center_x_lens_light2', 'center_y_lens_light2'
               ]

values = []
for i in range(n_params):
    values.append([median_pos[i][last - 1],
                   q84_pos[i][last - 1] - median_pos[i][last - 1],
                   median_pos[i][last - 1] - q16_pos[i][last - 1]])

param_mcmc_dicts = dict(zip(list_params, values))

#filtered by list params
parameter_values = {k: param_mcmc_dicts.get(k) for k in list_params}

# Write the dictionary to a YAML file
yaml_file_name = "DESIJ1624+0129_point_estimates.yml"
with open(yaml_file_name, 'w') as yaml_file:
    # Convert NumPy arrays to lists before writing to YAML
    parameter_values_list = {k: v.tolist() if isinstance(v, np.ndarray) else v for k, v in parameter_values.items()}
    yaml.dump(parameter_values_list, yaml_file, default_flow_style=False)

print(parameter_values_list)
print(f"YAML file '{yaml_file_name}' created successfully.")

```

It was created successfully. But later, I tried to read the file. It's showing an error, which hasn't been solved yet. Do you have any suggestions, @ajshajib vai?
![WhatsApp Image 2023-11-28 at 19 12 31_fc73be48](https://github.com/ajshajib/BDLensing/assets/78560474/06f66812-cf6a-4de9-b848-916d3d6c7889)

**Reply by ajshajib** [reply]
I don't have any suggestions at this point except for Googling about it. Can you open the file with a text editor by clicking it?

**Reply by aaimtiaz** [reply]
Yes. It's opening with VSCode. I'll further work on it and let you know the updates by tomorrow.

**Comment by aaimtiaz** [comment]
Problem solved. Now it's working properly.

```
import yaml

# list of parameters to be included in the yaml file
list_params = ['theta_E_lens0', 'gamma_lens0',
               'e1_lens0', 'e2_lens0',
               'center_x_lens0', 'center_y_lens0',
               'gamma_ext_lens1', 'psi_ext_lens1',
               'theta_E_lens2', 'R_sersic_lens_light0',
               'e1_lens_light0', 'e2_lens_light0',
               'center_x_lens_light0', 'center_y_lens_light0',
               'R_sersic_lens_light1', 'e1_lens_light1',
               'e2_lens_light1', 'R_sersic_lens_light2',
               'e1_lens_light2', 'e2_lens_light2',
               'center_x_lens_light2', 'center_y_lens_light2'
               ]

# values of the parameters from mcmc stored in a 2D array
values = []
for i in range(n_params):
    values.append([float(median_pos[i][last - 1]),
                   float(q84_pos[i][last - 1] - median_pos[i][last - 1]),
                   float(median_pos[i][last - 1] - q16_pos[i][last - 1])])

# making a dict of mcmc parameters
param_mcmc_dicts = dict(zip(param_mcmc, values))

# parameters filtered by list params
parameter_values = {k: param_mcmc_dicts.get(k) for k in list_params}

# Write the dictionary to a YAML file
yaml_file_name = "DESIJ1624+0129_point_estimates.yml"
with open(yaml_file_name, 'w') as yaml_file:
    yaml.dump(parameter_values, yaml_file, default_flow_style=False)
print(f"YAML file '{yaml_file_name}' created successfully.")
```

**Comment by SulymanHossain** [comment]
Should we add the Modeling History part as a requirement for the final deliverables? Especially for the paper writer, this information would be really helpful. Another option is to create new Issues for this task, which might delay this further. What do you suggest? @ajshajib

**Reply by ajshajib** [reply]
It would be nice to have it as part of the final deliverables, but I am afraid of making it a requirement if it slows things down. So, making separate issues to be resolved sometime in the future would be safer, I think.
